engines/tantivy-0.20/src/lib.rs

```diff
@@ -24,8 +24,10 @@ mod whitespace_tokenizer_fork {
     }
 
     impl Tokenizer for WhitespaceTokenizer {
-        fn token_stream<'a>(&self, text: &'a str) -> BoxTokenStream<'a> {
-            BoxTokenStream::from(WhitespaceTokenStream {
+        type TokenStream<'a> = BoxTokenStreamFork<'a>;
```

Comment:
here you could say

```
type TokenStream<'a> = WhitespaceTokenStream<'a>
```

and construct a `WhitespaceTokenStream`. We don't need to Box it anymore

Reply:
Thanks will do this instead.
I was thinking of this but I wanted to stay as close to the original code as possible.

Can I ask, how come we no longer need to Box it?

Reply:
Oh boy -- let me give my best to explain this. 

In 0.19 the Tokenizer trait looks like this 

```rust
pub trait Tokenizer: 'static + Send + Sync + TokenizerClone {
    /// Creates a token stream for a given `str`.
    fn token_stream<'a>(&self, text: &'a str) -> BoxTokenStream<'a>;
}

```rust

and in 0.20 it has changed to 
```
pub trait Tokenizer: 'static + Clone + Send + Sync {
    /// The token stream returned by this Tokenizer.
    type TokenStream<'a>: TokenStream;
    /// Creates a token stream for a given `str`.
    fn token_stream<'a>(&'a mut self, text: &'a str) -> Self::TokenStream<'a>;
}
```

what's interesting is that instead of saying every Tokenizer implementation should give back a `BoxTokenStream<'a>` (just ignore the <'a> for now), we let the `Tokenizer` impl for a type `Foo` to make a decision of what kind of TokenStream<'a> it implements. 

An analogy will be the Iterator trait which lets the implementor decides what kind of Item a type can iterate. For example, we have a type called Foo which internally holds a byte array. We can implement `Iterator<Item=i32>` and `Iterator<Item=u32>` or others for Foo by interpreting the bytes differently. The `Item` type with the Iterator trait definition is called **associated type**

```rust
pub trait Iterator {
    /// The type of the elements being iterated over.
    type Item;
...
```

The Iterator trait example is straightforward. Going back to Tokenizer -- the associated type `TokenStream<a'>` has a generic component which rust couldn't support until [1.65](https://blog.rust-lang.org/2022/11/03/Rust-1.65.0.html#generic-associated-types-gats) with the feature called GAT (generic associated types)

So clearly tantivy 0.20 made use of it and relaxed the requirement of Tokenizer implementors.